EpsilonCCS/Src/CanParsers/MotorCanParser.c

```diff
@@ -0,0 +1,252 @@
+            motor1DetailsData.railSupply3_3v = arrayToFloat(data[4]);
+            break;
+
+        case M1_HEATSINK_TEMPERATURE_CAN_ID:
+            motor1DetailsData.motorTemperature = arrayToFloat(data[0]);
+            motor1DetailsData.heatSinkTemperature = arrayToFloat(data[4]);
+            break;
+
+        case M1_DSP_BOARD_TEMPERATURE_CAN_ID:
+            motor1DetailsData.dspBoardTemp = arrayToFloat(data[0]);
+            break;
+
+        case M1_ODOMETER_BUS_AMPHOURS_CAN_ID:
+            motor1DetailsData.odometer = arrayToFloat(data[0]);
+            motor1DetailsData.dcBusAmpHours = arrayToFloat(data[4]);
+            break;
+
+        case M1_SLIP_SPEED_CAN_ID:
+            motor1DetailsData.slipSpeed = arrayToFloat(data[4]);
+            break;
+    }
+}
+
+void parseMotorFaultMessage(uint8_t motorId, uint8_t* data)
+{
+    struct MotorErrorFlags* errorData;
+    struct MotorLimitFlags* limitData;
+    unsigned char* rxErrorData;
+    unsigned char* txErrorData;
+
+    if (motorId == M0_ID)
+    {
+        errorData = &motorFaultsData.m0ErrorFlags;
+        limitData = &motorFaultsData.m0LimitFlags;
+        rxErrorData = &motorFaultsData.m0CanRxErrorCount;
+        txErrorData = &motorFaultsData.m0CanTxErrorCount;
+    }
+    else if (motorId == M1_ID)
+    {
+        errorData = &motorFaultsData.m1ErrorFlags;
+        limitData = &motorFaultsData.m1LimitFlags;
+        rxErrorData = &motorFaultsData.m1CanRxErrorCount;
+        txErrorData = &motorFaultsData.m1CanTxErrorCount;
+    }
+    else
+    {
+        return;
+    }
+
+    (*limitData).outputVoltagePwm =
+        data[0] & OUTPUT_VOLTAGE_PWM_LIMIT_MASK;
+    (*limitData).motorCurrent =
+        data[0] & MOTOR_CURRENT_LIMIT_MASK;
+    (*limitData).velocity =
+        data[0] & VELOCITY_LIMIT_MASK;
+    (*limitData).busCurrent =
+        data[0] & BUS_CURRENT_LIMIT_MASK;
+    (*limitData).busVoltageUpper =
+        data[0] & BUS_VOLTAGE_UPPER_LIMIT_MASK;
+    (*limitData).busVoltageLower =
+        data[0] & BUS_VOLTAGE_LOWER_LIMIT_MASK;
+    (*limitData).ipmOrMotorTemperature =
+        data[0] & IPM_OR_MOTOR_TEMPERATURE_LIMIT_MASK;
+
+    (*errorData).motorOverSpeed =
+        data[1] & MOTOR_OVER_SPEED_MASK;
+    (*errorData).softwareOverCurrent =
+        data[0] & SOFTWARE_OVER_CURRENT_MASK;
+    (*errorData).dcBusOverVoltage =
+        data[0] & DC_BUS_OVER_VOLTAGE_MASK;
+    (*errorData).badMotorPositionHallSequence =
+        data[0] & BAD_MOTOR_POSITION_HALL_SEQUENCE_MASK;
+    (*errorData).watchdogCausedLastReset =
+        data[0] & WATCHDOG_CAUSED_LAST_RESET_MASK;
+    (*errorData).configReadError =
+        data[0] & CONFIG_READ_ERROR_MASK;
+    (*errorData).rail15vUnderVoltageLockOut =
+        data[0] & RAIL_15V_UNDER_VOLTAGE_LOCK_OUT_MASK;
+    (*errorData).desaturationFault =
+        data[0] & DESATURATION_FAULT_MASK;
+
+    *rxErrorData = data[7];
+    *txErrorData = data[6];
+}
+
+float arrayToFloat(uint8_t* data)
+{
+    float result =
```

Comment:
I this won't work as you are just assigning the integer representation to the float. I.E if you have a floating point number of 16.0, in hex that is 0x41800000. Now that x86 is little endian the internal contents of the float need to end up as 0x00008041 and not just equal to 32833. I would suggest going the union approach from the mbed version of CCS as that is a pretty standard way of converting and much cleaner. If you are opposed to using unions/don't want to use any of the old CCS code then here is another way:

```C
#include <stdio.h>

int main() 
{
   // https://www.h-schmidt.net/FloatConverter/IEEE754.html
   unsigned char data[4] = {0x41, 0x80, 0x00, 0x00}; // 16.0
   float result = 0;
   float resultEnoch = (data[0] << 0) |
      (data[1] << 8) |
      (data[2] << 16) |
      (data[3] << 24);
   
   ((unsigned char*)&result)[0] = data[3];
   ((unsigned char*)&result)[1] = data[2];
   ((unsigned char*)&result)[2] = data[1];
   ((unsigned char*)&result)[3] = data[0];

   printf("result is %f, Enoch method %f\n", result, resultEnoch);
   return 0;
}
```

This outputs
```
jorda00@DESTINY:/mnt/e/Users/jordan/workspace/solarcarTest$ ./a.out
result is 16.000000, Enoch method 32833.000000
```